pkg/pipeline/pipeline.go

```diff
@@ -149,31 +156,81 @@ func NewPipeline() (*Pipeline, error) {
 		Writer:       writer,
 	}
 
+	// TODO: all this apply* functions can be removed if we change the Ingester, Transformer, etc...
+	// interfaces and use them directly here
+	ingests := node.AsInit(p.Ingester.Ingest)
+	decodes := node.AsMiddle(p.applyDecode)
+	transforms := node.AsMiddle(p.applyTransforms)
+	writes := node.AsTerminal(p.applyWrite)
+	extracts := node.AsMiddle(p.applyExtract)
+	encodes := node.AsTerminal(p.applyEncode)
+
+	ingests.SendsTo(decodes)
+	decodes.SendsTo(transforms)
+	transforms.SendsTo(writes, extracts)
+	extracts.SendsTo(encodes)
+
+	p.start = []*node.Init{ingests}
+	p.terminal = []*node.Terminal{writes, encodes}
+
 	return p, nil
 }
 
 func (p *Pipeline) Run() {
 	p.IsRunning = true
-	p.Ingester.Ingest(p.Process)
+	p.process()
 	p.IsRunning = false
 }
 
-// Process is called by the Ingester function
-func (p Pipeline) Process(entries []interface{}) {
+// Process is builds the processing graph and waits until it finishes
+func (p Pipeline) process() {
 	log.Debugf("entering pipeline.Process")
-	log.Debugf("number of entries = %d", len(entries))
-	decoded := p.Decoder.Decode(entries)
-	transformed := make([]config.GenericMap, 0)
-	var flowEntry config.GenericMap
-	for _, entry := range decoded {
-		flowEntry = transform.ExecuteTransforms(p.Transformers, entry)
-		transformed = append(transformed, flowEntry)
+
+	// starting the graph
+	for _, s := range p.start {
+		s.Start()
+	}
+
+	// blocking the execution until the graph terminal stages to end before
+	for _, t := range p.terminal {
+		<-t.Done()
+	}
```

Comment:
No, this jus stops the execution of this goroutine until the pipeline has ended (which at the moment would be the same as "wait forever", since we haven't yet implemented any mechanism co cancel the execution).